File: agent/schemas/estate.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


AssetType = Literal[
    "real_estate",
    "bank_account",
    "retirement",
    "vehicle",
    "personal_property",
    "other",
]
DebtType = Literal["secured", "unsecured", "priority"]
AlertSeverity = Literal["critical", "warning", "info"]
AlertType = Literal["deadline", "liability", "missing_doc", "rule_violation"]
AlertTimingStatus = Literal["dated", "blocking", "prerequisite", "missing_data", "no_deadline"]
TaskStatus = Literal["todo", "in_progress", "done", "blocked"]
EstatePhase = Literal[1, 2, 3, 4, 5, 6]


class ContractModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Executor(ContractModel):
    name: str
    email: str


class Asset(ContractModel):
    id: str
    type: AssetType
    description: str
    estimatedValue: float | None = None
    appraised: bool = False
    appraisedValue: float | None = None
    beneficiaryNamed: bool | None = None


class Debt(ContractModel):
    id: str
    creditor: str
    amount: float
    type: DebtType
    notified: bool = False
    notifiedDate: str | None = None
    claimFiled: bool | None = None


class Beneficiary(ContractModel):
    id: str
    name: str
    share: str | None = None
    specificBequest: str | None = None
    contactInfo: str | None = None


class UploadedDocument(ContractModel):
    id: str
    fileName: str
    documentType: str
    uploadedAt: str = Field(default_factory=utc_now_iso)
    source: str | None = None


class Task(ContractModel):
    id: str
    title: str
    status: TaskStatus = "todo"
    phase: EstatePhase
    dueDate: str | None = None
    relatedAlertId: str | None = None


class Alert(ContractModel):
    id: str
    severity: AlertSeverity
    type: AlertType
    title: str
    body: str
    rule: str
    daysRemaining: int | None = None
    timingStatus: AlertTimingStatus = "no_deadline"
    actionRequired: str
    whatYouNeed: list[str] = Field(default_factory=list)
    steps: list[str] = Field(default_factory=list)
    createdAt: str = Field(default_factory=utc_now_iso)
    dismissed: bool = False


class SavedLetter(ContractModel):
    id: str
    letterType: str
    recipientName: str | None = None
    draft: str
    savedAt: str = Field(default_factory=utc_now_iso)


class EstateState(ContractModel):
    id: str
    deceasedName: str
    dateOfDeath: str
    appointmentDate: str
    state: Literal["california"] = "california"
    county: str | None = None
    executor: Executor
    assets: list[Asset] = Field(default_factory=list)
    debts: list[Debt] = Field(default_factory=list)
    beneficiaries: list[Beneficiary] = Field(default_factory=list)
    documents: list[UploadedDocument] = Field(default_factory=list)
    tasks: list[Task] = Field(default_factory=list)
    alerts: list[Alert] = Field(default_factory=list)
    letters: list[SavedLetter] = Field(default_factory=list)
    phase: EstatePhase = 1
    createdAt: str = Field(default_factory=utc_now_iso)
    updatedAt: str = Field(default_factory=utc_now_iso)

File: web/types/estate.ts
export type AssetType =
  | "real_estate"
  | "bank_account"
  | "retirement"
  | "vehicle"
  | "personal_property"
  | "other";

export type AlertSeverity = "critical" | "warning" | "info";
export type AlertType = "deadline" | "liability" | "missing_doc" | "rule_violation";
export type AlertTimingStatus = "dated" | "blocking" | "prerequisite" | "missing_data" | "no_deadline";
export type EstatePhase = 1 | 2 | 3 | 4 | 5 | 6;
export type TaskStatus = "todo" | "in_progress" | "done" | "blocked";

export interface Executor {
  name: string;
  email: string;
}

export interface Asset {
  id: string;
  type: AssetType;
  description: string;
  estimatedValue?: number | null;
  appraised: boolean;
  appraisedValue?: number | null;
  beneficiaryNamed?: boolean | null;
}

export interface Debt {
  id: string;
  creditor: string;
  amount: number;
  type: "secured" | "unsecured" | "priority";
  notified: boolean;
  notifiedDate?: string | null;
  claimFiled?: boolean | null;
}

export interface Beneficiary {
  id: string;
  name: string;
  share?: string | null;
  specificBequest?: string | null;
  contactInfo?: string | null;
}

export interface UploadedDocument {
  id: string;
  fileName: string;
  documentType: string;
  uploadedAt: string;
  source?: string | null;
}

export interface Task {
  id: string;
  title: string;
  status: TaskStatus;
  phase: EstatePhase;
  dueDate?: string | null;
  relatedAlertId?: string | null;
}

export interface Alert {
  id: string;
  severity: AlertSeverity;
  type: AlertType;
  title: string;
  body: string;
  rule: string;
  daysRemaining?: number | null;
  timingStatus?: AlertTimingStatus;
  actionRequired: string;
  whatYouNeed: string[];
  steps: string[];
  createdAt: string;
  dismissed: boolean;
}

export interface SavedLetter {
  id: string;
  letterType: string;
  recipientName?: string | null;
  draft: string;
  savedAt: string;
}

export interface EstateState {
  id: string;
  deceasedName: string;
  dateOfDeath: string;
  appointmentDate: string;
  state: "california";
  county?: string | null;
  executor: Executor;
  assets: Asset[];
  debts: Debt[];
  beneficiaries: Beneficiary[];
  documents: UploadedDocument[];
  tasks: Task[];
  alerts: Alert[];
  letters: SavedLetter[];
  phase: EstatePhase;
  createdAt: string;
  updatedAt: string;
}
